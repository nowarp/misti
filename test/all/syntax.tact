//
// Checks if the tools process various Tact syntax constructions.
//

trait T {
  abstract get fun getter(): Int;
  abstract const something: Int;
}
contract TestContract with T {
  override get fun getter(): Int { return 0 }
  override const something: Int = 42;
}
trait Z {
  a: Int;
  get fun getter(): Int {
      return self.a + 1;
  }
}
contract TestContractF {
    a: Int = 42;
    fun test() {}
}
trait TParent {
  a: Int;
  fun getA(): Int {
      return self.a;
  }
}
trait TChild with TParent {
  a: Int; // Declare 'a' to satisfy TParent's requirement
}
contract TestContractT with TChild {
    a: Int = 19;
    fun test(): Int { return self.getA(); }
    external() {}
}

fun test_try(a: Int) {
    try { a += 1; }
    try { /* empty */ }
    try { a = 19; } catch (err) { dump(err); }
}

fun test_loops() {
    let sum: Int = 0;
    let i: Int = 0;
    while (i < 10) {
      i = i + 1;
      sum = sum + i;
    }
    do {
      i = i - 1;
      sum = sum + i;
    } until (i > 10);
    repeat (10) {
      i = i + 1;
      sum = sum + i;
    }
    let m: map<Int, Int> = emptyMap();
    foreach (k, v in m) {}
}

fun testTryCatch(a: Int) {
    try {
        a += 1;
    } catch (err) {
        dump(err);
    }
}

fun testLoops() {
    let sum: Int = 0;
    let i: Int = 0;

    while (i < 10) {
        i = i + 1;
        sum = sum + i;
    }

    do {
        i = i - 1;
        sum = sum + i;
    } until (i <= 0);

    repeat (10) {
        i = i + 1;
        sum = sum + i;
    }

    let m: map<Int, Int> = emptyMap();
    m.set(1, 10);
    m.set(2, 20);
    m.set(3, 30);

    foreach (key, value in m) {
        sum = sum + value;
    }
}

// Test callgraph effects
contract EffectTest {
    addr: Address;
    // Effect.StateWrite
    init() { self.addr = sender(); }

    // Effect.Send + Effect.StateRead
    fun funcWithSend() {
        let amount: Int = 100;
        send(SendParameters{ to: self.addr, value: amount });
    }

    // Effect.StateRead
    fun funcWithStateRead() {
        let value = self.addr;
    }

    // Effect.StateWrite
    fun funcWithStateWrite() {
        self.addr = newAddress(0, 0);
    }

    // Multiple effects:
    // Effect.AccessDatetime + Effect.PrgUse + Effect.SeedInit
    fun funcWithMultipleEffects() {
        let currentTime: Int = now();
        let randValue: Int = random(1, 100);
        if (1 > 42) {
            nativeRandomizeLt();
        }
    }
}
