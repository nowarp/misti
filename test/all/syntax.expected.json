{
  "kind": "warnings",
  "warnings": [
    {
      "detectorId": "NeverAccessedVariables",
      "description": "Variable value is never accessed",
      "location": {
        "file": "test/all/syntax.tact",
        "line": 125,
        "column": 9,
        "code": "  124 |     fun funcWithStateRead() {\n> 125 |         let value = self.addr;\n                ^~~~~~~~~~~~~~~~~~~~~~\n  126 |     }\n"
      },
      "suppressed": false,
      "severity": 3,
      "category": 1,
      "docURL": "https://nowarp.io/tools/misti/docs/detectors/NeverAccessedVariables",
      "suggestion": "Consider removing the variable",
      "quickfixes": []
    },
    {
      "detectorId": "NeverAccessedVariables",
      "description": "Variable currentTime is never accessed",
      "location": {
        "file": "test/all/syntax.tact",
        "line": 141,
        "column": 9,
        "code": "  140 |     fun funcWithMultipleEffects() {\n> 141 |         let currentTime: Int = now();\n                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n  142 |         let randValue: Int = random(1, 100);\n"
      },
      "suppressed": false,
      "severity": 3,
      "category": 1,
      "docURL": "https://nowarp.io/tools/misti/docs/detectors/NeverAccessedVariables",
      "suggestion": "Consider removing the variable",
      "quickfixes": []
    },
    {
      "detectorId": "NeverAccessedVariables",
      "description": "Variable randValue is never accessed",
      "location": {
        "file": "test/all/syntax.tact",
        "line": 142,
        "column": 9,
        "code": "  141 |         let currentTime: Int = now();\n> 142 |         let randValue: Int = random(1, 100);\n                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n  143 |         if (1 > 42) {\n"
      },
      "suppressed": false,
      "severity": 3,
      "category": 1,
      "docURL": "https://nowarp.io/tools/misti/docs/detectors/NeverAccessedVariables",
      "suggestion": "Consider removing the variable",
      "quickfixes": []
    },
    {
      "detectorId": "FalseCondition",
      "description": "Condition always evaluates to false",
      "location": {
        "file": "test/all/syntax.tact",
        "line": 143,
        "column": 9,
        "code": "  142 |         let randValue: Int = random(1, 100);\n> 143 |         if (1 > 42) {\n                ^~~~~~~~~~~~~\n  144 |             nativeRandomizeLt();\n"
      },
      "suppressed": false,
      "severity": 3,
      "category": 1,
      "docURL": "https://nowarp.io/tools/misti/docs/detectors/FalseCondition",
      "quickfixes": [
        {
          "kind": "delete",
          "description": "Remove the condition",
          "shown": true,
          "range": {
            "start": {
              "line": 143,
              "column": 9
            },
            "end": {
              "line": 3,
              "column": 10
            }
          }
        }
      ]
    },
    {
      "detectorId": "FalseCondition",
      "description": "Condition always evaluates to false",
      "location": {
        "file": "test/all/syntax.tact",
        "line": 196,
        "column": 9,
        "code": "  195 |         while (false) {}\n> 196 |         while (false) {\n                ^~~~~~~~~~~~~~~\n  197 |             dump(\"something\");\n"
      },
      "suppressed": false,
      "severity": 3,
      "category": 1,
      "docURL": "https://nowarp.io/tools/misti/docs/detectors/FalseCondition",
      "quickfixes": [
        {
          "kind": "delete",
          "description": "Remove the condition",
          "shown": true,
          "range": {
            "start": {
              "line": 196,
              "column": 9
            },
            "end": {
              "line": 3,
              "column": 10
            }
          }
        }
      ]
    },
    {
      "detectorId": "FalseCondition",
      "description": "Condition always evaluates to false",
      "location": {
        "file": "test/all/syntax.tact",
        "line": 200,
        "column": 9,
        "code": "  199 |         // Suggested to replace code\n> 200 |         if (1 == 2) { return true; }\n                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~\n  201 |         else { return false; }\n"
      },
      "suppressed": false,
      "severity": 3,
      "category": 1,
      "docURL": "https://nowarp.io/tools/misti/docs/detectors/FalseCondition",
      "quickfixes": [
        {
          "kind": "delete",
          "description": "Remove the condition",
          "shown": true,
          "range": {
            "start": {
              "line": 200,
              "column": 9
            },
            "end": {
              "line": 2,
              "column": 31
            }
          }
        }
      ]
    },
    {
      "detectorId": "SuspiciousLoop",
      "description": "Infinite loop detected",
      "location": {
        "file": "test/all/syntax.tact",
        "line": 166,
        "column": 16,
        "code": "  165 |     receive() {\n> 166 |         while (true) {\n                       ^~~~\n  167 |             self.m.set(1, 2);\n"
      },
      "suppressed": false,
      "severity": 3,
      "category": 1,
      "docURL": "https://nowarp.io/tools/misti/docs/detectors/SuspiciousLoop",
      "suggestion": "Avoid unbounded conditions in loops",
      "quickfixes": []
    },
    {
      "detectorId": "SuspiciousLoop",
      "description": "Loop condition is always false",
      "location": {
        "file": "test/all/syntax.tact",
        "line": 196,
        "column": 16,
        "code": "  195 |         while (false) {}\n> 196 |         while (false) {\n                       ^~~~~\n  197 |             dump(\"something\");\n"
      },
      "suppressed": false,
      "severity": 3,
      "category": 1,
      "docURL": "https://nowarp.io/tools/misti/docs/detectors/SuspiciousLoop",
      "suggestion": "The condition is always false; the body will never execute",
      "quickfixes": []
    },
    {
      "detectorId": "SuboptimalSend",
      "description": "Prefer `message` over `send`",
      "location": {
        "file": "test/all/syntax.tact",
        "line": 120,
        "column": 9,
        "code": "  119 |         let amount: Int = 100;\n> 120 |         send(SendParameters{ to: self.addr, value: amount });\n                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n  121 |     }\n"
      },
      "suppressed": false,
      "severity": 3,
      "category": 2,
      "docURL": "https://nowarp.io/tools/misti/docs/detectors/SuboptimalSend",
      "suggestion": "Use more gas-efficient `message` function: https://docs.tact-lang.org/ref/core-common/#message",
      "quickfixes": []
    },
    {
      "detectorId": "ZeroAddress",
      "description": "Using zero address",
      "location": {
        "file": "test/all/syntax.tact",
        "line": 130,
        "column": 35,
        "code": "  129 |     fun funcWithStateWrite1() {\n> 130 |         self.addr = newAddress(0, 0);\n                                          ^\n  131 |         self.b.storeAddress(self.addr);\n"
      },
      "suppressed": false,
      "severity": 2,
      "category": 1,
      "docURL": "https://nowarp.io/tools/misti/docs/detectors/ZeroAddress",
      "suggestion": "Consider changing code to avoid using it. For example, you could pass the address during the deployment.",
      "quickfixes": []
    },
    {
      "detectorId": "UnboundMap",
      "description": "Map self.m could be unbound",
      "location": {
        "file": "test/all/syntax.tact",
        "line": 161,
        "column": 5,
        "code": "  160 | contract A {\n> 161 |     m: map<Int, Int> = null;\n            ^~~~~~~~~~~~~~~~~~~~~~~\n  162 |     get fun m(): map<Int, Int> {\n"
      },
      "suppressed": false,
      "severity": 2,
      "category": 1,
      "extraDescription": "There are operations adding elements to this map, but there is no API to remove them",
      "docURL": "https://nowarp.io/tools/misti/docs/detectors/UnboundMap",
      "suggestion": "Consider adding a method to remove elements or suppress this warning",
      "quickfixes": []
    },
    {
      "detectorId": "EtaLikeSimplifications",
      "description": "Return the condition directly",
      "location": {
        "file": "test/all/syntax.tact",
        "line": 200,
        "column": 9,
        "code": "  199 |         // Suggested to replace code\n> 200 |         if (1 == 2) { return true; }\n                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~\n  201 |         else { return false; }\n"
      },
      "suppressed": false,
      "severity": 2,
      "category": 2,
      "docURL": "https://nowarp.io/tools/misti/docs/detectors/EtaLikeSimplifications",
      "quickfixes": [
        {
          "kind": "replace",
          "shown": true,
          "description": "Return the condition directly",
          "replacements": [
            {
              "range": {
                "start": {
                  "line": 200,
                  "column": 9
                },
                "end": {
                  "line": 2,
                  "column": 31
                }
              },
              "value": "return 1 == 2;"
            }
          ]
        }
      ]
    },
    {
      "detectorId": "DumpIsUsed",
      "description": "Found `dump` usage",
      "location": {
        "file": "test/all/syntax.tact",
        "line": 41,
        "column": 35,
        "code": "  40 |     try { /* empty */ }\n> 41 |     try { a = 19; } catch (err) { dump(err); }\n                                         ^~~~~~~~~\n  42 | }\n"
      },
      "suppressed": false,
      "severity": 1,
      "category": 1,
      "extraDescription": "Using `dump` in production code can sometimes indicate complex code that requires additional review",
      "docURL": "https://nowarp.io/tools/misti/docs/detectors/DumpIsUsed",
      "quickfixes": []
    },
    {
      "detectorId": "DumpIsUsed",
      "description": "Found `dump` usage",
      "location": {
        "file": "test/all/syntax.tact",
        "line": 67,
        "column": 9,
        "code": "  66 |     } catch (err) {\n> 67 |         dump(err);\n               ^~~~~~~~~\n  68 |     }\n"
      },
      "suppressed": false,
      "severity": 1,
      "category": 1,
      "extraDescription": "Using `dump` in production code can sometimes indicate complex code that requires additional review",
      "docURL": "https://nowarp.io/tools/misti/docs/detectors/DumpIsUsed",
      "quickfixes": []
    },
    {
      "detectorId": "PreferAugmentedAssign",
      "description": "Prefer augmented assignment: i += 1",
      "location": {
        "file": "test/all/syntax.tact",
        "line": 48,
        "column": 7,
        "code": "  47 |     while (i < 10) {\n> 48 |       i = i + 1;\n             ^~~~~~~~~~\n  49 |       sum = sum + i;\n"
      },
      "suppressed": false,
      "severity": 1,
      "category": 3,
      "docURL": "https://nowarp.io/tools/misti/docs/detectors/PreferAugmentedAssign",
      "suggestion": "Consider using augmented assignment instead: i += 1",
      "quickfixes": []
    },
    {
      "detectorId": "PreferAugmentedAssign",
      "description": "Prefer augmented assignment: sum += i",
      "location": {
        "file": "test/all/syntax.tact",
        "line": 49,
        "column": 7,
        "code": "  48 |       i = i + 1;\n> 49 |       sum = sum + i;\n             ^~~~~~~~~~~~~~\n  50 |     }\n"
      },
      "suppressed": false,
      "severity": 1,
      "category": 3,
      "docURL": "https://nowarp.io/tools/misti/docs/detectors/PreferAugmentedAssign",
      "suggestion": "Consider using augmented assignment instead: sum += i",
      "quickfixes": []
    },
    {
      "detectorId": "PreferAugmentedAssign",
      "description": "Prefer augmented assignment: i -= 1",
      "location": {
        "file": "test/all/syntax.tact",
        "line": 52,
        "column": 7,
        "code": "  51 |     do {\n> 52 |       i = i - 1;\n             ^~~~~~~~~~\n  53 |       sum = sum + i;\n"
      },
      "suppressed": false,
      "severity": 1,
      "category": 3,
      "docURL": "https://nowarp.io/tools/misti/docs/detectors/PreferAugmentedAssign",
      "suggestion": "Consider using augmented assignment instead: i -= 1",
      "quickfixes": []
    },
    {
      "detectorId": "PreferAugmentedAssign",
      "description": "Prefer augmented assignment: sum += i",
      "location": {
        "file": "test/all/syntax.tact",
        "line": 53,
        "column": 7,
        "code": "  52 |       i = i - 1;\n> 53 |       sum = sum + i;\n             ^~~~~~~~~~~~~~\n  54 |     } until (i > 10);\n"
      },
      "suppressed": false,
      "severity": 1,
      "category": 3,
      "docURL": "https://nowarp.io/tools/misti/docs/detectors/PreferAugmentedAssign",
      "suggestion": "Consider using augmented assignment instead: sum += i",
      "quickfixes": []
    },
    {
      "detectorId": "PreferAugmentedAssign",
      "description": "Prefer augmented assignment: i += 1",
      "location": {
        "file": "test/all/syntax.tact",
        "line": 56,
        "column": 7,
        "code": "  55 |     repeat (10) {\n> 56 |       i = i + 1;\n             ^~~~~~~~~~\n  57 |       sum = sum + i;\n"
      },
      "suppressed": false,
      "severity": 1,
      "category": 3,
      "docURL": "https://nowarp.io/tools/misti/docs/detectors/PreferAugmentedAssign",
      "suggestion": "Consider using augmented assignment instead: i += 1",
      "quickfixes": []
    },
    {
      "detectorId": "PreferAugmentedAssign",
      "description": "Prefer augmented assignment: sum += i",
      "location": {
        "file": "test/all/syntax.tact",
        "line": 57,
        "column": 7,
        "code": "  56 |       i = i + 1;\n> 57 |       sum = sum + i;\n             ^~~~~~~~~~~~~~\n  58 |     }\n"
      },
      "suppressed": false,
      "severity": 1,
      "category": 3,
      "docURL": "https://nowarp.io/tools/misti/docs/detectors/PreferAugmentedAssign",
      "suggestion": "Consider using augmented assignment instead: sum += i",
      "quickfixes": []
    },
    {
      "detectorId": "PreferAugmentedAssign",
      "description": "Prefer augmented assignment: i += 1",
      "location": {
        "file": "test/all/syntax.tact",
        "line": 76,
        "column": 9,
        "code": "  75 |     while (i < 10) {\n> 76 |         i = i + 1;\n               ^~~~~~~~~~\n  77 |         sum = sum + i;\n"
      },
      "suppressed": false,
      "severity": 1,
      "category": 3,
      "docURL": "https://nowarp.io/tools/misti/docs/detectors/PreferAugmentedAssign",
      "suggestion": "Consider using augmented assignment instead: i += 1",
      "quickfixes": []
    },
    {
      "detectorId": "PreferAugmentedAssign",
      "description": "Prefer augmented assignment: sum += i",
      "location": {
        "file": "test/all/syntax.tact",
        "line": 77,
        "column": 9,
        "code": "  76 |         i = i + 1;\n> 77 |         sum = sum + i;\n               ^~~~~~~~~~~~~~\n  78 |     }\n"
      },
      "suppressed": false,
      "severity": 1,
      "category": 3,
      "docURL": "https://nowarp.io/tools/misti/docs/detectors/PreferAugmentedAssign",
      "suggestion": "Consider using augmented assignment instead: sum += i",
      "quickfixes": []
    },
    {
      "detectorId": "PreferAugmentedAssign",
      "description": "Prefer augmented assignment: i -= 1",
      "location": {
        "file": "test/all/syntax.tact",
        "line": 81,
        "column": 9,
        "code": "  80 |     do {\n> 81 |         i = i - 1;\n               ^~~~~~~~~~\n  82 |         sum = sum + i;\n"
      },
      "suppressed": false,
      "severity": 1,
      "category": 3,
      "docURL": "https://nowarp.io/tools/misti/docs/detectors/PreferAugmentedAssign",
      "suggestion": "Consider using augmented assignment instead: i -= 1",
      "quickfixes": []
    },
    {
      "detectorId": "PreferAugmentedAssign",
      "description": "Prefer augmented assignment: sum += i",
      "location": {
        "file": "test/all/syntax.tact",
        "line": 82,
        "column": 9,
        "code": "  81 |         i = i - 1;\n> 82 |         sum = sum + i;\n               ^~~~~~~~~~~~~~\n  83 |     } until (i <= 0);\n"
      },
      "suppressed": false,
      "severity": 1,
      "category": 3,
      "docURL": "https://nowarp.io/tools/misti/docs/detectors/PreferAugmentedAssign",
      "suggestion": "Consider using augmented assignment instead: sum += i",
      "quickfixes": []
    },
    {
      "detectorId": "PreferAugmentedAssign",
      "description": "Prefer augmented assignment: i += 1",
      "location": {
        "file": "test/all/syntax.tact",
        "line": 86,
        "column": 9,
        "code": "  85 |     repeat (10) {\n> 86 |         i = i + 1;\n               ^~~~~~~~~~\n  87 |         sum = sum + i;\n"
      },
      "suppressed": false,
      "severity": 1,
      "category": 3,
      "docURL": "https://nowarp.io/tools/misti/docs/detectors/PreferAugmentedAssign",
      "suggestion": "Consider using augmented assignment instead: i += 1",
      "quickfixes": []
    },
    {
      "detectorId": "PreferAugmentedAssign",
      "description": "Prefer augmented assignment: sum += i",
      "location": {
        "file": "test/all/syntax.tact",
        "line": 87,
        "column": 9,
        "code": "  86 |         i = i + 1;\n> 87 |         sum = sum + i;\n               ^~~~~~~~~~~~~~\n  88 |     }\n"
      },
      "suppressed": false,
      "severity": 1,
      "category": 3,
      "docURL": "https://nowarp.io/tools/misti/docs/detectors/PreferAugmentedAssign",
      "suggestion": "Consider using augmented assignment instead: sum += i",
      "quickfixes": []
    },
    {
      "detectorId": "PreferAugmentedAssign",
      "description": "Prefer augmented assignment: sum += value",
      "location": {
        "file": "test/all/syntax.tact",
        "line": 96,
        "column": 9,
        "code": "  95 |     foreach (key, value in m) {\n> 96 |         sum = sum + value;\n               ^~~~~~~~~~~~~~~~~~\n  97 |     }\n"
      },
      "suppressed": false,
      "severity": 1,
      "category": 3,
      "docURL": "https://nowarp.io/tools/misti/docs/detectors/PreferAugmentedAssign",
      "suggestion": "Consider using augmented assignment instead: sum += value",
      "quickfixes": []
    },
    {
      "detectorId": "AsmIsUsed",
      "description": "asm function is used",
      "location": {
        "file": "test/all/syntax.tact",
        "line": 156,
        "column": 1,
        "code": "  155 | // Checks if we add asm functions to IR entries correctly\n> 156 | asm(-> 1 0) extends mutates fun loadRefEx(self: Slice): Cell { LDREF }\n        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n  157 | fun callAsm(s: Slice): Cell { return s.loadRefEx(); }\n"
      },
      "suppressed": false,
      "severity": 1,
      "category": 1,
      "extraDescription": "Using TVM assembly is a potentially dangerous operation that requires additional review",
      "docURL": "https://nowarp.io/tools/misti/docs/detectors/AsmIsUsed",
      "quickfixes": []
    }
  ],
  "logs": {
    "debug": [
      "Enabled detectors (40): DivideBeforeMultiply, ReadOnlyVariables, NeverAccessedVariables, UnboundLoop, ZeroAddress, ConstantAddress, BranchDuplicate, DumpIsUsed, FieldDoubleInit, PreferAugmentedAssign, StringReceiversOverlap, AsmIsUsed, PreferredStdlibApi, InheritedStateMutation, ArgCopyMutation, OptimalMathFunction, DuplicatedCondition, UnusedOptional, EnsurePrgSeed, FalseCondition, SendInLoop, UnboundMap, StateMutationInGetter, UnusedExpressionResult, SuspiciousMessageMode, ShortCircuitCondition, EtaLikeSimplifications, ExitCodeUsage, CellBounds, UnprotectedCall, SuspiciousLoop, SuboptimalSend, PreferBinaryReceiver, PreferSenderFunction, ImplicitOpcode, SuboptimalCellOperation, DuplicatedImport, TransitiveImport, RedundantBooleanExpression, UnusedMethodArgument",
      "Enabled tools (0): ",
      "Parsing project syntax ...",
      "Parsing completed: syntax",
      "node_modules/@tact-lang/compiler/dist/stdlib/stdlib/std/internal/text.tact:   424 |     let b: StringBuilder = beginComment();\n> 425 |     b.append(self);\n            ^\n  426 |     return b.toCell();\n\nUnsupported contract method access: id",
      "node_modules/@tact-lang/compiler/dist/stdlib/stdlib/std/internal/text.tact:   524 | extends fun fromBase64(self: Slice): Slice {\n> 525 |     let size: Int = self.bits() / 8;\n                            ^~~~\n  526 |     let result: Builder = beginCell();\n\nAccessing an unknown contract: self",
      "node_modules/@tact-lang/compiler/dist/stdlib/stdlib/std/internal/text.tact:   528 |     repeat (size) {\n> 529 |         let code: Int = self.loadUint(8);\n                                ^~~~\n  530 |         if (code >= 65 && code <= 90) { // A-Z\n\nAccessing an unknown contract: self",
      "node_modules/@tact-lang/compiler/dist/stdlib/stdlib/std/internal/text.tact:   530 |         if (code >= 65 && code <= 90) { // A-Z\n> 531 |             result = result.storeUint(code - 65, 6);\n                             ^~~~~~\n  532 |         } else if (code >= 97 && code <= 122) { // a-z\n\nUnsupported contract method access: id",
      "node_modules/@tact-lang/compiler/dist/stdlib/stdlib/std/internal/text.tact:   532 |         } else if (code >= 97 && code <= 122) { // a-z\n> 533 |             result = result.storeUint(code - (97 - 26), 6);\n                             ^~~~~~\n  534 |         } else if (code >= 48 && code <= 57) { // 0-9\n\nUnsupported contract method access: id",
      "node_modules/@tact-lang/compiler/dist/stdlib/stdlib/std/internal/text.tact:   534 |         } else if (code >= 48 && code <= 57) { // 0-9\n> 535 |             result = result.storeUint(code + (52 - 48), 6);\n                             ^~~~~~\n  536 |         } else if (code == 45 || code == 43) { // - or +\n\nUnsupported contract method access: id",
      "node_modules/@tact-lang/compiler/dist/stdlib/stdlib/std/internal/text.tact:   536 |         } else if (code == 45 || code == 43) { // - or +\n> 537 |             result = result.storeUint(62, 6);\n                             ^~~~~~\n  538 |         } else if (code == 95 || code == 47) { // _ or /\n\nUnsupported contract method access: id",
      "node_modules/@tact-lang/compiler/dist/stdlib/stdlib/std/internal/text.tact:   538 |         } else if (code == 95 || code == 47) { // _ or /\n> 539 |             result = result.storeUint(63, 6);\n                             ^~~~~~\n  540 |         } else if (code == 61) { // =\n\nUnsupported contract method access: id",
      "node_modules/@tact-lang/compiler/dist/stdlib/stdlib/std/internal/text.tact:   547 |     // Padding\n> 548 |     let total: Int = result.bits();\n                             ^~~~~~\n  549 |     let padding: Int = total % 8;\n\nUnsupported contract method access: id",
      "node_modules/@tact-lang/compiler/dist/stdlib/stdlib/std/internal/text.tact:   550 |     if (padding != 0) {\n> 551 |         let s: Slice = result.asSlice();\n                               ^~~~~~\n  552 |         return s.loadBits(total - padding);\n\nUnsupported contract method access: id",
      "node_modules/@tact-lang/compiler/dist/stdlib/stdlib/std/internal/contract.tact:   141 | inline extends fun hasSameBasechainAddress(self: StateInit, address: Address): Bool {\n> 142 |     let addressHash = parseStdAddress(address.asSlice()).address;\n                                              ^~~~~~~\n  143 |     let baseAddress = contractBasechainAddress(self);\n\nUnsupported contract method access: id",
      "node_modules/@tact-lang/compiler/dist/stdlib/stdlib/std/internal/send.tact:   582 |     //                     maybe: stateInit (false) bodyRef: bool (true)\n> 583 |     let c: Cell = beginCell()\n                          ^~~~~~~~~~~\n  584 |         .storeUint(15211807202738752817960438464513, 104)\n\nUnsupported contract method access: method_call",
      "node_modules/@tact-lang/compiler/dist/stdlib/stdlib/std/internal/send.tact:   582 |     //                     maybe: stateInit (false) bodyRef: bool (true)\n> 583 |     let c: Cell = beginCell()\n                          ^~~~~~~~~~~\n  584 |         .storeUint(15211807202738752817960438464513, 104)\n\nUnsupported contract method access: method_call",
      "node_modules/@tact-lang/compiler/dist/stdlib/stdlib/std/internal/send.tact:   582 |     //                     maybe: stateInit (false) bodyRef: bool (true)\n> 583 |     let c: Cell = beginCell()\n                          ^~~~~~~~~~~\n  584 |         .storeUint(15211807202738752817960438464513, 104)\n\nUnsupported contract method access: static_call",
      "node_modules/@tact-lang/compiler/dist/stdlib/stdlib/std/internal/address.tact:   28 |     // 256 bits for the address itself\n> 29 |     throwUnless(TactExitCodeInvalidStandardAddress, self.bits() == 267);\n                                                           ^~~~\n  30 | \n\nAccessing an unknown contract: self",
      "node_modules/@tact-lang/compiler/dist/stdlib/stdlib/std/internal/address.tact:   33 |         // i.e. internal address prefix and chain ID -1\n> 34 |         throwUnless(TactExitCodeInvalidStandardAddress, self.preloadUint(11) == 1279);\n                                                               ^~~~\n  35 |     } else {\n\nAccessing an unknown contract: self",
      "node_modules/@tact-lang/compiler/dist/stdlib/stdlib/std/internal/address.tact:   37 |         // but do not verify the chain ID\n> 38 |         throwUnless(TactExitCodeInvalidStandardAddress, self.preloadUint(3) == 4);\n                                                               ^~~~\n  39 |     }\n\nAccessing an unknown contract: self",
      "test/all/syntax.tact:   90 |     let m: map<Int, Int> = emptyMap();\n> 91 |     m.set(1, 10);\n           ^\n  92 |     m.set(2, 20);\n\nUnsupported contract method access: id",
      "test/all/syntax.tact:   91 |     m.set(1, 10);\n> 92 |     m.set(2, 20);\n           ^\n  93 |     m.set(3, 30);\n\nUnsupported contract method access: id",
      "test/all/syntax.tact:   92 |     m.set(2, 20);\n> 93 |     m.set(3, 30);\n           ^\n  94 | \n\nUnsupported contract method access: id",
      "test/all/syntax.tact:   173 | extends fun toInt(self: String): Int {\n> 174 |     let string: Slice = self.asSlice();\n                                ^~~~\n  175 |     let acc: Int = 0;\n\nAccessing an unknown contract: self",
      "test/all/syntax.tact:   176 |     while (!string.empty()) {\n> 177 |         let char: Int = string.loadUint(8);\n                                ^~~~~~\n  178 |         acc = (acc * 10) + (char - 48);\n\nUnsupported contract method access: id",
      "test/all/syntax.tact:   130 |         self.addr = newAddress(0, 0);\n> 131 |         self.b.storeAddress(self.addr);\n                ^~~~~~\n  132 |         self.s.b.storeAddress(self.addr);\n\nUnsupported contract method access: field_access",
      "test/all/syntax.tact:   131 |         self.b.storeAddress(self.addr);\n> 132 |         self.s.b.storeAddress(self.addr);\n                ^~~~~~~~\n  133 |     }\n\nUnsupported contract method access: field_access",
      "test/all/syntax.tact:   150 |         let a = beginString();\n> 151 |         a.append(\"f\");\n                ^\n  152 |     }\n\nUnsupported contract method access: id",
      "test/all/syntax.tact:   166 |         while (true) {\n> 167 |             self.m.set(1, 2);\n                    ^~~~~~\n  168 |         }\n\nUnsupported contract method access: field_access",
      "CGNode created without AST ID for function \"dump\"",
      "CGNode created without AST ID for function \"emptyMap\"",
      "CGNode created without AST ID for function \"m::set\"",
      "CGNode created without AST ID for function \"s::loadRefEx\"",
      "CGNode created without AST ID for function \"String::asSlice\"",
      "CGNode created without AST ID for function \"string::empty\"",
      "CGNode created without AST ID for function \"string::loadUint\"",
      "CGNode created without AST ID for function \"TestContractT::getA\"",
      "CGNode created without AST ID for function \"sender\"",
      "CGNode created without AST ID for function \"beginCell\"",
      "CGNode created without AST ID for function \"send\"",
      "CGNode created without AST ID for function \"s::loadAddress\"",
      "CGNode created without AST ID for function \"now\"",
      "CGNode created without AST ID for function \"nativeRandomizeLt\"",
      "CGNode created without AST ID for function \"beginString\"",
      "CGNode created without AST ID for function \"a::append\""
    ],
    "info": [],
    "warn": [],
    "error": []
  }
}